File: C/dp/houseRobber.c
#include <stdio.h>
#include <stdlib.h>

int rob(int num[], int n) {
    if (n == 0) return 0;
    int *d = (int *)calloc(n + 2, sizeof(int));
    int i;
    d[0] = d[1] = 0; 
    for (i = 2; i < n + 2; i++) {
        if (d[i - 2] + num[i - 2] > d[i - 1]) {
            d[i] = d[i - 2] + num[i - 2];
        }
        else {
            d[i] = d[i - 1];
        }
    }
    return d[n + 1];
}

int main() {
    int num;
    printf("Enter number of houses: ");
    scanf("%d",&num);
    int money[num];
    printf("\nEnter money in each house: ");
    for(int i=0;i<num;i++)
    {
           scanf("%d",&money[i]);
    }
    printf("\nTotal amount that can be robbed: %d", rob(money, num));

    return 0;
}

/*
Time Complexity: n(log(n))
Space Complexity: n
Sample input: 
Enter number of houses: 4
Enter money in each house: 1 2 3 1 
Total amount that can be robbed: 4
*/

File: Python/dp/House_Robber.py
#A thief wants to loot houses. He knows the amount of money in each house. He cannot loot two consecutive houses. Find the maximum amount of money he can loot.


def maxMoneyLooted(amountInHouses, totalHouses) :
    #If there are 0 houses , return 0 
    if totalHouses == 0 :
        return 0

    #If there is only one house , return its money    
    if totalHouses == 1 :
        return amountInHouses[0]

    #If there are atleast 2 houses ,create arr and calculate the maximum amount of money that can be looted
    maxMoneyCanBeLooted = [0]*totalHouses

    maxMoneyCanBeLooted[0] = amountInHouses[0]
    maxMoneyCanBeLooted[1] = max(amountInHouses[0], amountInHouses[1])

    for currHouse in range(2,totalHouses) :
        #At each i th house either (houses upto i-2) + (i th house) or (houses upto i-1) can be robbed
        maxMoneyCanBeLooted[currHouse] = max(maxMoneyCanBeLooted[currHouse - 1], (amountInHouses[currHouse] + maxMoneyCanBeLooted[currHouse - 2]))

    #Return the required answer which is the maximum amount of money that can be robbed till last house    
    return maxMoneyCanBeLooted[totalHouses-1]


def takeInput() :
    n = int(input().rstrip())
    if n == 0 :
        return list(), 0

    arr = list(map(int, input().rstrip().split(" ")))
    return arr, n

def main():
    print('Enter total number of houses and amount that each house contains: ')
    amountInHouses,totalHouses = takeInput()
    print('Maximum amount that can be taken be the thief is: ')
    print(maxMoneyLooted(amountInHouses,totalHouses))

if __name__ == '__main__':
    main()


'''
Time Complexity: O(n)
Space Complexity: O(n)
where n is the total number of houses

Sample Input:
Enter total number of houses and amount that each house contains:
6
10 2 30 20 3 50

Sample Output:
Maximum amount that can be taken be the thief is:
90
Explanation :
Looting first, third, and the last houses([10 + 30 + 50]) will result in the maximum loot,
and all the other possible combinations would result in less than 90.
'''
